firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java

```diff
@@ -0,0 +1,347 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.core;
+
+import static com.google.firebase.firestore.util.Assert.hardAssert;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import com.google.firebase.firestore.Filter;
+import com.google.firebase.firestore.model.Document;
+import com.google.firestore.v1.StructuredQuery;
+import com.google.firestore.v1.StructuredQuery.CompositeFilter.Operator;
+import java.util.ArrayList;
+import java.util.List;
+
+interface FieldFilterCondition {
+  boolean check(FieldFilter field);
+}
+
+/** Represents a filter that is the conjunction or disjunction of other filters. */
+public class CompositeFilter extends Filter {
+  private final List<Filter> filters;
+  //  private final boolean isAnd;
+  private final Operator operator;
+
+  public CompositeFilter(List<Filter> filters, Operator operator) {
+    this.filters = filters;
+    this.operator = operator;
+  }
+
+  public List<Filter> getFilters() {
+    return filters;
+  }
+
+  public Operator getOperator() {
+    return operator;
+  }
+
+  public boolean isAnd() {
+    return operator == Operator.AND;
+  }
+
+  public boolean isOr() {
+    // TODO(ehsann): Replace with Operator.OR.
+    return operator == Operator.OPERATOR_UNSPECIFIED;
+  }
+
+  /**
+   * Returns true if this is an AND filter and all the filters in this composite filter are
+   * FieldFilters (i.e. it does not contain any other composite filters).
+   */
+  public boolean isFlatAndFilter() {
+    return isAnd() && !containsCompositeFilters();
+  }
+
+  /**
+   * Returns true if this composite filter contains any other composite filter. Returns false
+   * otherwise.
+   */
+  public boolean containsCompositeFilters() {
+    for (Filter filter : filters) {
+      if (filter instanceof CompositeFilter) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Performs a depth-first search to find and return the first FieldFilter in the composite filter
+   * that satisfies the condition. Returns null if none of the FieldFilters satisfy the condition.
+   */
+  public FieldFilter firstFieldFilterWhere(FieldFilterCondition condition) {
+    for (Filter filter : filters) {
+      if (filter instanceof FieldFilter && condition.check(((FieldFilter) filter))) {
+        return (FieldFilter) filter;
+      } else if (filter instanceof CompositeFilter) {
```

Comment:
Can we change this logic so that it does not have to do an instanceof check? Every Filter could have a `firstFieldFilterWhere` style method. Most will just return `this` or `null`.